crypto/src/swap.rs

```diff
@@ -26,10 +56,129 @@ pub struct SwapPlaintext {
 }
 
 impl SwapPlaintext {
-    // TODO: needs to be constant-length
-    pub fn encrypt(&self, _ovk: &OutgoingViewingKey) -> SwapCiphertext {
-        // TODO: implement
-        todo!()
+    // Create a new hash based on the ephemeral public key and shared secret suitable for use as a key for symmetric encryption.
+    //
+    // Implementing this way allows recovery of all swap plaintexts via the seed phrase.
+    //
+    // Theoretically, if a paranoid user did want to achieve forward secrecy, they could choose to encrypt
+    // nonsense bytes as the swap plaintext as the swap ciphertext does not need to be valid for the
+    // swap to succeed, however this is unsupported by the official client.
+    fn derive_symmetric_key(
+        shared_secret: &ka::SharedSecret,
+        epk: &ka::Public,
+    ) -> blake2b_simd::Hash {
+        let mut kdf_params = blake2b_simd::Params::new();
+        kdf_params.hash_length(32);
+        let mut kdf = kdf_params.to_state();
+        kdf.update(&shared_secret.0);
+        kdf.update(&epk.0);
+
+        kdf.finalize()
+    }
+
+    pub fn diversified_generator(&self) -> decaf377::Element {
+        self.b_d
+    }
+
+    pub fn transmission_key(&self) -> ka::Public {
+        self.pk_d
+    }
+
+    /// Use Blake2b-256 to derive an encryption key `ock` from the OVK and public fields.
+    pub fn derive_ock(ovk: &OutgoingViewingKey, epk: &ka::Public) -> blake2b_simd::Hash {
+        // let cv_bytes: [u8; 32] = cv.into();
+        // let cm_bytes: [u8; 32] = cm.into();
+
+        let mut kdf_params = blake2b_simd::Params::new();
+        kdf_params.hash_length(32);
+        let mut kdf = kdf_params.to_state();
+        kdf.update(&ovk.0);
+        // TODO: should we be using the public fields e.g. t1, t2, trading_pair here?
+        // Note implementation uses value commitments...
+        // kdf.update(&cv_bytes);
+        // kdf.update(&cm_bytes);
+        kdf.update(&epk.0);
+
+        kdf.finalize()
+    }
```

Comment:
We don't currently have a distinct ock type, do we?

Since the goal is to have the swap plaintexts be encrypted with the same key as the note plaintext (so that any entity that can see the note recording the swap NFT can also see the swap data it commits to, and vice versa), I think it would be good to find a way to share code between note encryption and swap plaintext encryption.

I wonder whether creating a common OCK type would give us a convenient place to put that code?

Reply:
cc @redshiftzero

Reply:
We do not have an Ock type, that seems like a good idea though (could file a followup for it if we wanted to out of scope for this PR). For this PR we do have the `note::derive_ock` function that could be shared - I think we can use the value and note commitment bytes of the swap note plaintext to derive the OCK key